hive3/src/main/java/org/apache/iceberg/mr/hive/vector/ParquetSchemaFieldNameVisitor.java

```diff
@@ -55,17 +58,25 @@ public Type struct(Types.StructType expected, GroupType struct, List<Type> field
 
     for (Types.NestedField field : expectedFields) {
       int id = field.fieldId();
-      if (id != MetadataColumns.ROW_POSITION.fieldId() && id != MetadataColumns.IS_DELETED.fieldId()) {
-        Type fieldInFileSchema = typesById.get(id);
-        if (fieldInFileSchema == null) {
-          // New field - not in this parquet file yet, add the new field name instead of null
+      if (id == MetadataColumns.ROW_POSITION.fieldId() || id == MetadataColumns.IS_DELETED.fieldId()) {
+        continue;
+      }
+      Type fieldInPrunedFileSchema = typesById.get(id);
+      if (fieldInPrunedFileSchema == null) {
+        if (!originalFileSchema.containsField(field.name())) {
+          // Must be a new field - it isn't in this parquet file yet, so add the new field name instead of null
           appendToColNamesList(isMessageType, field.name());
         } else {
-          // Already present column in this parquet file, add the original name
-          types.add(fieldInFileSchema);
-          appendToColNamesList(isMessageType, fieldInFileSchema.getName());
+          // This field is found in the parquet file with a different ID, so it must have been recreated since.
+          // Inserting a dummy col name to force Hive Parquet reader returning null for this column.
+          appendToColNamesList(isMessageType, DUMMY_COL_NAME);
```

Comment:
What happens if we have to such column? Is it ok with parquet if we provide duplicate column names?

Reply:
If we have what column? If we have found a column which is not in the pruned schema, but it is in the file schema, and their names match then it must be a column that was dropped and then re-added. If that's the case we add this dummy column so that the reader returns nulls.

Reply:
to -> two 😄

Reply:
Got it ;) Good point actually, I checked this case and the reader luckily isn't bothered by seeing more of the (same) dummy names in the column list. For each such dummy, null values are returned to Hive which is correct behaviour.